ICU-6065 UnicodeSet::closeOver(simple case folding)

markusicu opened the pull request:
Adds a new option for UnicodeSet::closeOver(int options) and other functions with this options parameter:
- C (unicode/uset.h) USET_SIMPLE_CASE_INSENSITIVE
- Java UnicodeSet.SIMPLE_CASE_INSENSITIVE

With the new option, the UnicodeSet is expanded so that for each code point the other code points in its Simple_Case_Folding (scf) equivalence class are added, and each multi-character string is replaced with a version where each code point is replaced with its scf mapping.

The implementation uses the same data structure as for the full case folding (UnicodeSet.CASE_INSENSITIVE) but only adds the “simple” (single-code point) mappings, and also omits a hardcoded list of mappings where there is no scf equivalence. (I will propose additional scf mappings in a future version of Unicode to fix those.)

There is a new test function which computes the scf equivalence classes and compares closeOver(SIMPLE_CASE_INSENSITIVE) with applying that test-code closure data.

As an attempted optimization, when the UnicodeSet is large, the closeOver() code only looks at the code points with the Case_Sensitive property. It looks at fewer code points at the cost of lazy-initializing the Case_Insensitive set and computing the intersection.

There are other changes for some code cleanup and minor refactoring. For example, in the UnicodeSet implementation, there are now two separate functions for CASE_INSENSITIVE/SIMPLE_CASE_INSENSITIVE vs. ADD_CASE_MAPPINGS. In the low-level code, the hardcoded mappings for letters like ‘i’ are moved into the slower case-exception-handling path. The Java closeOver() test had been missing some test data that the C++ test had.

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-6065
- [x] Required: The PR title must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. <!-- For example: "ICU-1234 Fix xyz" -->
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

macchiati reviewed (commented):
Looks nice!

markusicu commented:
@macchiati @richgillam please take a look at the "better docs" commit here.

FYI: C++ is complete (modulo feedback) but I have not yet started the Java port, so I am keeping this PR draft+incomplete.

richgillam reviewed (approved):
The doc changes look good to me.

markusicu commented:
I think I finished the Java port. It should be complete. Please review, again probably best one commit at a time. I will look it over myself once more as well. I will squash after addressing feedback.

markusicu commented on icu4c/source/common/characterproperties.cpp:
```diff
@@ -377,22 +377,30 @@ UCPMap *makeMap(UProperty property, UErrorCode &errorCode) {
-U_CAPI const USet * U_EXPORT2
-u_getBinaryPropertySet(UProperty property, UErrorCode *pErrorCode) {
-    if (U_FAILURE(*pErrorCode)) { return nullptr; }
+const UnicodeSet *CharacterProperties::getBinaryPropertySet(UProperty property, UErrorCode &errorCode) {
```
FYI  This is just creating an internal C++ API which is easier to use (less casting) from C++ impl code than the public C API.

markusicu reviewed (commented):
I looked over the changes myself and made very minor docs & test changes. I think I am done. Please review :-)

markusicu commented:
I wrote a lengthy description for this pull request. Hope this helps with reviewing as well as documentation for posterity.

richgillam reviewed (dismissed):
The API and documentation changes look good to me.  I don't think I'm qualified to review the implementation changes, but as far as I could figure them out, they looked good too.

macchiati commented on icu4j/main/classes/core/src/com/ibm/icu/text/UnicodeSet.java:
```diff
@@ -3861,58 +3869,81 @@ private void applyPropertyPattern(RuleCharacterIterator chars,
+     */
+    public static final int SIMPLE_CASE_INSENSITIVE = 6;
+
+    private static final int CASE_MASK = CASE_INSENSITIVE | ADD_CASE_MAPPINGS;
```
Question: would it be useful to note which of these don't make sense in combination? eg what would 0xFF do?

macchiati replied:
Or are they really just singletons?

markusicu replied:
In the public-API docs, I documented
- the valid options
- that the case-related ones are mutually exclusive
- (in a couple of places) that undefined bits are ignored/reserved

I don't think we need to add more on this internal constant.

macchiati replied:
ok, no problem.

macchiati commented on icu4j/main/classes/core/src/com/ibm/icu/text/UnicodeSet.java:
```diff
@@ -3930,97 +3961,185 @@ private static final void addCaseMapping(UnicodeSet set, int result, StringBuild
+        if (src.size() < 30) {
+            return src;
+        }
+        return CharacterProperties.getBinaryPropertySet(UProperty.CASE_SENSITIVE).
```
Not a blocker.
Would it be faster to clone the src and then retainAll CharacterProperties.getBinaryPropertySet(UProperty.CASE_SENSITIVE), because the src is likely to be smaller? Or maybe clone the smaller of the two?

markusicu replied:
Hard to say without a benchmark. It's probably cheaper to clone the one with fewer *ranges*, and if the src.hasStrings() then cloning src unnecessarily clones the structure for that too.

Do you want me to do something like
```
UnicodeSet sensitive = ...
if (src.hasStrings() || src.getRangeCount() > sensitive.getRangeCount()) {
    return sensitive.cloneAsThawed().retainAll(src);
} else {
    return src.clone().retainAll(sensitive);
}
```
?

macchiati replied:
You have a good point. Probably not worth doing without a benchmark.

markusicu replied:
I just did this in the latest commit.

macchiati commented:
I didn't review the C++, just the java. I had no objections — just wanted to get a read from you on some of my comments.

markusicu commented on icu4j/main/classes/core/src/com/ibm/icu/text/UnicodeSet.java:
```diff
@@ -3861,58 +3869,81 @@ private void applyPropertyPattern(RuleCharacterIterator chars,
      *
+     * @stable ICU 3.8
+     */
+    public static final int CASE = 2;
```
could do --> https://unicode-org.atlassian.net/browse/ICU-22299

macchiati replied:
great, thanks.

markusicu commented on icu4j/main/tests/core/src/com/ibm/icu/dev/test/lang/UnicodeSetTest.java:
```diff
@@ -1323,38 +1325,98 @@ public void TestScriptNames() {
 
+                SIMPLE_CASE_INSENSITIVE,
+                "[aq\u00DF{Bc}{bC}{Fi}]",
+                "[aAqQ\u00DF\u1E9E{bc}{fi}]",
```
I can see if I can make a quick change to that effect. I would probably keep combining marks, and hard-to-distinguish characters like Kelvin, escaped.

markusicu replied:
I tried this, and I don't like it. Half of the characters want to remain unescaped as noted above. But for several of the others it's also more obfuscating what's going on than helpful.

macchiati replied:
I agree with combining marks, di, controls, etc. 

Not sure about combining marks or confusables. I find 
```
"[\u0390]",          
"[\u0390\u1FD3{\u03B9\u0308\u0301}]", 

```
more opaque than
```
"[ΐ]",
"[ΐ{ΐ}ΐ]",
```

But maybe one or the other in comments?

markusicu commented:
Thanks for reviewing! I just squashed, please re-approve.
